maven-caching/src/main/java/org/apache/maven/caching/CacheControllerImpl.java

```diff
@@ -397,67 +415,126 @@ private RestoredArtifact restoredArtifact( org.apache.maven.artifact.Artifact pa
         return artifact;
     }
 
-    private Future<File> createDownloadTask( CacheResult cacheResult, CacheContext context, MavenProject project,
-                                             Artifact artifact )
-            throws ExecutionException, InterruptedException
+    //we might store in cache artifact which was build with previous version
+    //1.0-SNAPSHOT is kept in cache but real version of project is 2.0
+    //for pom packaging this is done automatically by maven but for jar and other there might be
+    //sensitive metadata with previous version. Versions mismatch could lead errors
+    private Path adjustArchiveArtifactVersion( MavenProject project, String originalArtifactVersion, Path artifactFile )
+            throws IOException
     {
-        final FutureTask<File> downloadTask = new FutureTask<>( () -> 
+
+        File file = artifactFile.toFile();
+        if ( project.getVersion().equals( originalArtifactVersion ) || !isArchive( file ) )
         {
-            LOGGER.debug( "Downloading artifact {}", artifact.getArtifactId() );
-            final Path artifactFile = localCache.getArtifactFile( context, cacheResult.getSource(),
-                    artifact );
-            if ( !Files.exists( artifactFile ) )
+            return artifactFile;
+        }
+
+        String currentVersion = project.getVersion();
+        File tmpJarFile = File.createTempFile( artifactFile.toFile().getName(),
```

Comment:
i woul better avois using tmp - on many machines tmp could be 2G or similar. All downloaded artifacts might just not fit there.  There is an [File.html#createTempFile](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#createTempFile(java.lang.String,%20java.lang.String,%20java.io.File)) method which takes directory as a parameter - it is better to use it and create artifacts somethere under the target dir (in assumption that there should be enough space to create them) + easy to clean.

Reply:
> 

should we avoid potential directory names clash? how to do this?

Reply:
Risk of conflicts is not higher than in any over cases for any over plugins. I don't even know if need to solve this problem. But artifacts need to stay for a time of the build so everyone of them need to fit tmp in order to succeed. It might be not convenient.

Reply:
> To be clear, i didn't meant to restore artifacts on original paths. I meant create temp files in a project 'target/build-cache-tmp'. Risk of conflicts in thta case is not higher than in any over cases for any over plugin. I don't even know if need to solve this problem. But in current design all artifacts need to stay for a time of the build so each of them needs to fit tmp in order to succeed. It might be not convenient and cause env specific failures.


changed destination to target/cache-build-tmp